### PR TITLE
Feature/integral line filtering

### PR DESCRIPTION
**Pull request by r-englund**
Moved my Line Filtering module from my own repo. 

image

### CODE REVIEW

**Comment by r-englund** [comment]
2 weeks old now, anyone feel like they wanna review this?

File: vectorvis/integrallinefiltering/include/inviwo/integrallinefiltering/utils/sparsehistogram.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+#include <inviwo/integrallinefiltering/integrallinefilteringmoduledefine.h>
+#include <inviwo/core/common/inviwo.h>
+
+namespace inviwo {
+
+/*
+ * A class to represent sparse histograms. Supports both signed and unsigned indices and multi
+ * dimensional histogram through the IndexType. For example, `SparseHistogram<int>` is a 1D
+ * histogram accepting both negative and positive indices, i.e. [INT_MIN INT_MAX].
```
Comment (martinfalk) [inline vectorvis/integrallinefiltering/include/inviwo/integrallinefiltering/utils/sparsehistogram.h]: should point out that it is only for integral types.

File: vectorvis/integrallinefiltering/readme.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Module that add processors and tools to allow the Integral Lines from the Vector Field Visualization module to be combined the the plotting and filtering from the Plotting modules.
```
Comment (martinfalk) [inline vectorvis/integrallinefiltering/readme.md]:
```
# IntegralLineFiltering Module 
Provides processors and tools for plotting and filtering Integral Lines from the Vector Field Visualization module. This module depends on the Plotting modules.
```

File: vectorvis/integrallinefiltering/src/processors/integrallinestodataframe.cpp [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+const std::string IntegralLinesToDataFrame::MetaDataProperty::classIdentifier =
+    "org.inviwo.IntegralLinesToDataFrame.MetaDataProperty";
+
+// The Class Identifier has to be globally unique. Use a reverse DNS naming scheme
+const ProcessorInfo IntegralLinesToDataFrame::processorInfo_{
+    "org.inviwo.IntegralLinesToDataFrame",  // Class identifier
+    "Integral Lines To Data Frame",         // Display name
+    "Integral Line Filtering",              // Category
+    CodeState::Experimental,                // Code state
+    Tags::None,                             // Tags
```
Comment (martinfalk) [inline vectorvis/integrallinefiltering/src/processors/integrallinestodataframe.cpp]: no tags?

File: vectorvis/integrallinefiltering/src/processors/integrallinestodataframe.cpp [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+    , includeTortuosity_("includeTurtuosity", "Include Tortuosity", true)
+    , includeTerminationReason_("includeTerminationReason", "Include Termination Reasons", true)
+    , includeEntropy_("includeEntropy", "Include Entropy", true)
+    , includeStartPositions_("includeStartPositions", "Include Line Start Coordinates", false)
+    , includeEndPositions_("includeEndPositions", "Include Line End Coordinates", false) {
+
+    addPort(lines_);
+    addPort(dataframe_);
+
+    addProperty(includeLineID_);
```
Comment (martinfalk) [inline vectorvis/integrallinefiltering/src/processors/integrallinestodataframe.cpp]: consider `addProperties`

File: vectorvis/integrallinefiltering/src/utils/directionalhistogram.cpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <inviwo/integrallinefiltering/utils/directionalhistogram.h>
+
+namespace inviwo {
+namespace {}  // namespace
```
Comment (martinfalk) [inline vectorvis/integrallinefiltering/src/utils/directionalhistogram.cpp]: remove

Comment (martinfalk) [inline vectorvis/integrallinefiltering/src/utils/directionalhistogram.cpp]: remove

File: vectorvis/integrallinefiltering/src/utils/shannonentropy.cpp [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <inviwo/integrallinefiltering/utils/shannonentropy.h>
+
+namespace inviwo {
+namespace util {}  // namespace util
```
Comment (martinfalk) [inline vectorvis/integrallinefiltering/src/utils/shannonentropy.cpp]: remove

**Comment by martinfalk** [comment]
@r-englund that's because there are >6000 lines to review ;-)

File: vectorvis/integrallinefiltering/include/inviwo/integrallinefiltering/utils/shannonentropy.h [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+ * @param numBins number of bins to use for the histogram
+ * @param normalize wether to use normalization or not, if normalization the output will be on the
+ * range [0 1], if not, the range will be on [0 log2(numBins)]
+ */
+template <size_t Dims, typename T>
+double shannonEntropyDirectional(const std::vector<glm::vec<Dims, T>>& data, const size_t numBins,
+                                 PerformNormalization normalize = PerformNormalization::Yes) {
+    static_assert(std::is_floating_point_v<T>);
+    static_assert(Dims == 2 || Dims == 3);
+    DirectionalHistogram<2, T> histogram(numBins);
```
[the remark this answers is not in the document]
Comment (r-englund) [reply vectorvis/integrallinefiltering/include/inviwo/integrallinefiltering/utils/shannonentropy.h]: Stupid GLM, the vec3 were implicit converted to vec2

File: vectorvis/integrallinefiltering/include/inviwo/integrallinefiltering/processors/integrallinestodataframe.h [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        BoolProperty w_{"w", "Use W-values", false};
+
+        BoolProperty log_{"log", "Log values", false};
+
+        BoolProperty avg_{"avg", "Include Average", true};
+        BoolProperty sd_{"sd", "Include Standard deviation", false};
+        StringProperty percentiles_{"percentiles",
+                                    "Percentiles (space separated, float [0-1] or ints (0-100) )"};
+
+        void initFunctions(std::vector<F> &funcs, const BufferRAM *ram, DataFrame &dataFrame);
```
[the remark this answers is not in the document]
Comment (r-englund) [reply vectorvis/integrallinefiltering/include/inviwo/integrallinefiltering/processors/integrallinestodataframe.h]: Maybe `MetaDataProperty` is a bad name. The `MetaDataProperty` contains settings for how/if line meta data should be included on the resulting dataframe. For example, mean velocity or mean curvature. 
I've renamed it to `MetaDataSettings` in a future commit

File: vectorvis/integrallinefiltering/include/inviwo/integrallinefiltering/utils/directionalhistogram.h [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+
+#include <inviwo/integrallinefiltering/integrallinefilteringmoduledefine.h>
+#include <inviwo/core/util/glm.h>
+#include <inviwo/core/util/foreach.h>
+
+#include <inviwo/integrallinefiltering/utils/sparsehistogram.h>
+
+namespace inviwo {
+
+namespace {
```
[the remark this answers is not in the document]
Comment (r-englund) [reply vectorvis/integrallinefiltering/include/inviwo/integrallinefiltering/utils/directionalhistogram.h]: yes, they are an implementation detail that need no infrmation from the class, therefore O opted to have them as stand alone functions. The option would be to have them as static private functions which I like less.

[the remark this answers is not in the document]
Comment (r-englund) [reply vectorvis/integrallinefiltering/include/inviwo/integrallinefiltering/utils/directionalhistogram.h]: I've renamed the functions to better reflect what they do and named the namespace detail

File: vectorvis/integrallinefiltering/src/processors/integrallinestodataframe.cpp (hunk 4 above)
Comment (r-englund) [reply vectorvis/integrallinefiltering/src/processors/integrallinestodataframe.cpp]: That function was not avaible when I implemented the processor ;)

File: vectorvis/integrallinefiltering/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+# Inviwo IntegralLineFiltering Module
+ivw_module(IntegralLineFiltering)
+
+#--------------------------------------------------------------------
+# Add header files
+set(HEADER_FILES
+    include/inviwo/integrallinefiltering/integrallinefilteringmodule.h
+    include/inviwo/integrallinefiltering/integrallinefilteringmoduledefine.h
+    include/inviwo/integrallinefiltering/processors/integrallinestodataframe.h
+    include/inviwo/integrallinefiltering/utils/directionalhistogram.h
```
Comment (petersteneteg) [inline vectorvis/integrallinefiltering/CMakeLists.txt]: We should try hard not to use util... this should probably be in datastructues
and the entropy in algorithm

File: vectorvis/integrallinefiltering/include/inviwo/integrallinefiltering/processors/integrallinestodataframe.h [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#ifndef IVW_INTEGRALLINESTODATAFRAME_H
```
Comment (petersteneteg) [inline vectorvis/integrallinefiltering/include/inviwo/integrallinefiltering/processors/integrallinestodataframe.h]: #pragma once

File: vectorvis/integrallinefiltering/include/inviwo/integrallinefiltering/processors/integrallinestodataframe.h [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+ *   * __Include Line Start Coordinates__ Check to include the {x,y,z} of the first vertex as a
+ * parameter.
+ *   * __Include Line End Coordinates__ Check to include {x,y,z} of the last vertex as a parameter.
+ *   * __Meta Data Settings__ Group of properties related meta data of the lines and wether or not
+ * to include it as a parameter.
+ *
+ */
+class IVW_MODULE_INTEGRALLINEFILTERING_API IntegralLinesToDataFrame : public Processor {
+public:
+    using F = std::function<void(const IntegralLine &line)>;
```
Comment (petersteneteg) [inline vectorvis/integrallinefiltering/include/inviwo/integrallinefiltering/processors/integrallinestodataframe.h]: better name?

**Comment by r-englund** [comment]
@petersteneteg Do you mind checking the last commit before merging this?  In that commit I introduce the standalone functions `calculateDirectionalHistogram(vectors,bins)` that computes the histograms.  I also introduced the class `UniformSpherePartitioning` that is basically what the `DirectionalHistogram<3, T>` did before, but without keeping the count per patch (it only determines which patch a point is within).

**Pull request merged**